### ISSUE TITLE
MySQL Error During Test Runs and Unit Tests on Travis CI

### ISSUE THREAD

**Issue by magneticstain** [magneticstain/Inquisition#66]
```
Traceback (most recent call last):
  File "/home/travis/build/magneticstain/Inquisition/lib/destiny/Erudite.py", line 404, in updateNodeOPSRecordInDB
    self.inquisitionDbHandle.commit()
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/pymysql/connections.py", line 788, in commit
    self._read_ok_packet()
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/pymysql/connections.py", line 769, in _read_ok_packet
    raise err.OperationalError(2014, "Command Out of Sync")
pymysql.err.OperationalError: (2014, 'Command Out of Sync')
```

**Comment by magneticstain** [comment]
The issue here is the same one documented at https://stackoverflow.com/q/32755572

Right now, all processes are using the same mysql connection generated from the parent process:
```
josh@jimmy1 ~/Code/Inquisition $ sudo netstat -ptan | grep 3306
tcp        0      0 127.0.0.1:3306          0.0.0.0:*               LISTEN      1198/mysqld     
tcp        0      0 127.0.0.1:59246         127.0.0.1:3306          ESTABLISHED 25123/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:59248         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:3306          127.0.0.1:59242         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:3306          127.0.0.1:59244         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:3306          127.0.0.1:59260         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:59250         127.0.0.1:3306          ESTABLISHED 25123/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:59252         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:59256         127.0.0.1:3306          ESTABLISHED 25123/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:59246         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:59252         127.0.0.1:3306          ESTABLISHED 25123/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:59256         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:59254         127.0.0.1:3306          ESTABLISHED 25123/python3   
tcp        0      0 127.0.0.1:59258         127.0.0.1:3306          ESTABLISHED 25123/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:59254         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:59248         127.0.0.1:3306          ESTABLISHED 25123/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:59250         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:3306          127.0.0.1:59258         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:59244         127.0.0.1:3306          ESTABLISHED 25123/python3   
tcp        0      0 127.0.0.1:59242         127.0.0.1:3306          ESTABLISHED 25123/python3   
tcp        0      0 127.0.0.1:59260         127.0.0.1:3306          ESTABLISHED 25123/python3   
josh@jimmy1 ~/Code/Inquisition $
```

I'm going to refactor the code in order to have each process utilize its own connection, as suggested in the SO article to maintain optimal performance.

**Comment by magneticstain** [comment]
Should have this fixed now. I removed Inquisit.generateInquisitionDbConnection() and replaced it with inquisit.bounceInquisitionDbConnection().
```
    def bounceInquisitionDbConnection(self):
        """
        Reconnects to MySQL for Inquisition DB connection

        :return: bool
        """

        # regenerate conn
        self.inquisitionDbHandle = pymysql.connect(host=self.cfg['mysql_database']['db_host'],
                                                   port=int(self.cfg['mysql_database']['db_port']),
                                                   user=self.cfg['mysql_database']['db_user'],
                                                   password=self.cfg['mysql_database']['db_pass'],
                                                   db=self.cfg['mysql_database']['db_name'],
                                                   charset='utf8mb4', cursorclass=pymysql.cursors.DictCursor)
        if self.inquisitionDbHandle:
            self.lgr.debug('database connection established for main inquisition database :: [ '
                           + self.cfg['mysql_database']['db_host'] + ':' + self.cfg['mysql_database']['db_port'] + ' ]')

            self.lgr.debug('all database connections established [ SUCCESSFULLY ]')

            return True
        else:
            return False
```
This function will reset the Inquisit class variable, bouncing the connection.

I tried to use the close() function in pymysql, but it actually generates an AttributeError exception for some reason:
```
Traceback (most recent call last):
  File "/home/josh/Code/Inquisition/inquisition.py", line 171, in <module>
    main()
  File "/home/josh/Code/Inquisition/inquisition.py", line 150, in main
    anatomize.startAnatomizer()
  File "/home/josh/Code/Inquisition/lib/anatomize/Anatomize.py", line 133, in startAnatomizer
    self.bounceInquisitionDbConnection()
  File "/home/josh/Code/Inquisition/lib/inquisit/Inquisit.py", line 116, in bounceInquisitionDbConnection
    self.inquisitionDbHandle.close()
  File "/usr/local/lib/python3.5/dist-packages/pymysql/connections.py", line 734, in close
    sock.close()
AttributeError: 'NoneType' object has no attribute 'close'
Traceback (most recent call last):
  File "/home/josh/Code/Inquisition/inquisition.py", line 171, in <module>
    main()
  File "/home/josh/Code/Inquisition/inquisition.py", line 150, in main
    anatomize.startAnatomizer()
  File "/home/josh/Code/Inquisition/lib/anatomize/Anatomize.py", line 133, in startAnatomizer
    self.bounceInquisitionDbConnection()
  File "/home/josh/Code/Inquisition/lib/inquisit/Inquisit.py", line 116, in bounceInquisitionDbConnection
    self.inquisitionDbHandle.close()
  File "/usr/local/lib/python3.5/dist-packages/pymysql/connections.py", line 734, in close
<pymysql.connections.Connection object at 0x7f20ed9d6358>
    sock.close()
AttributeError: 'NoneType' object has no attribute 'close'
```
This is not the first bug I've found with pymysql. I will create a separate issue report for tracking switching libraries.

**Comment by magneticstain** [comment]
We may actually need to close it if possible. We're generating a lot of TIME_WAIT connections when just overwriting the connection vars.

Might be hastening the switch documented in Issue #67.

```
josh@jimmy1 ~/Code/Inquisition $ sudo netstat -ptan | grep 3306
tcp        0      0 127.0.0.1:3306          0.0.0.0:*               LISTEN      1198/mysqld     
tcp        0      0 127.0.0.1:60938         127.0.0.1:3306          ESTABLISHED 32150/python3   
tcp        0      0 127.0.0.1:60934         127.0.0.1:3306          ESTABLISHED 32157/python3   
tcp        0      0 127.0.0.1:60668         127.0.0.1:3306          TIME_WAIT   -               
tcp        0      0 127.0.0.1:60942         127.0.0.1:3306          ESTABLISHED 32150/python3   
tcp        0      0 127.0.0.1:60674         127.0.0.1:3306          TIME_WAIT   -               
tcp        0      0 127.0.0.1:3306          127.0.0.1:60954         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60670         127.0.0.1:3306          TIME_WAIT   -               
tcp        0      0 127.0.0.1:60936         127.0.0.1:3306          ESTABLISHED 32150/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:60934         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60956         127.0.0.1:3306          ESTABLISHED 32152/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:60946         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60944         127.0.0.1:3306          ESTABLISHED 32150/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:60944         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60662         127.0.0.1:3306          TIME_WAIT   -               
tcp        0      0 127.0.0.1:60970         127.0.0.1:3306          ESTABLISHED 32160/python3   
tcp        0      0 127.0.0.1:60950         127.0.0.1:3306          ESTABLISHED 32150/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:60936         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:3306          127.0.0.1:60940         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60684         127.0.0.1:3306          TIME_WAIT   -               
tcp        0      0 127.0.0.1:60656         127.0.0.1:3306          TIME_WAIT   -               
tcp        0      0 127.0.0.1:3306          127.0.0.1:60960         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:3306          127.0.0.1:60966         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60954         127.0.0.1:3306          ESTABLISHED 32150/python3   
tcp        0      0 127.0.0.1:60940         127.0.0.1:3306          ESTABLISHED 32150/python3   
tcp        0      0 127.0.0.1:60948         127.0.0.1:3306          ESTABLISHED 32150/python3   
tcp        0      0 127.0.0.1:60660         127.0.0.1:3306          TIME_WAIT   -               
tcp        0      0 127.0.0.1:60678         127.0.0.1:3306          TIME_WAIT   -               
tcp        0      0 127.0.0.1:3306          127.0.0.1:60948         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60960         127.0.0.1:3306          ESTABLISHED 32154/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:60950         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:3306          127.0.0.1:60938         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60682         127.0.0.1:3306          TIME_WAIT   -               
tcp        0      0 127.0.0.1:3306          127.0.0.1:60970         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60666         127.0.0.1:3306          TIME_WAIT   -               
tcp        0      0 127.0.0.1:3306          127.0.0.1:60942         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60952         127.0.0.1:3306          ESTABLISHED 32150/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:60956         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60946         127.0.0.1:3306          ESTABLISHED 32150/python3   
tcp        0      0 127.0.0.1:3306          127.0.0.1:60964         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:3306          127.0.0.1:60952         ESTABLISHED 1198/mysqld     
tcp        0      0 127.0.0.1:60966         127.0.0.1:3306          ESTABLISHED 32156/python3   
tcp        0      0 127.0.0.1:60658         127.0.0.1:3306          TIME_WAIT   -               
tcp        0      0 127.0.0.1:60964         127.0.0.1:3306          ESTABLISHED 32157/python3   
josh@jimmy1 ~/Code/Inquisition $ 
```

**Comment by magneticstain** [comment]
https://sentry.io/carlsonet/inquisition/issues/407260256/

**Comment by magneticstain** [comment]
This has luckily been fixed with the latest commit - https://github.com/magneticstain/Inquisition/commit/9b3b18d7013ac1e6d17f32ffafdd1dad30c51667

I think with the issue only happening at startup and the probability of processes continuously starting and getting killed, we can wait until the next release to change MySQL libs.

**Comment by magneticstain** [comment]
Seems the connection issues are reoccurring in a critical fashion, but on travis CI:
```
ERROR: test_fetchParsers (test_anatomize.AnatomizeTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/magneticstain/Inquisition/lib/inquisit/Inquisit.py", line 61, in __init__
    self.bounceInquisitionDbConnection()
  File "/home/travis/build/magneticstain/Inquisition/lib/inquisit/Inquisit.py", line 119, in bounceInquisitionDbConnection
    charset='utf8mb4', cursorclass=pymysql.cursors.DictCursor)
  File "/home/travis/virtualenv/python3.5.3/lib/python3.5/site-packages/pymysql/__init__.py", line 90, in Connect
    return Connection(*args, **kwargs)
  File "/home/travis/virtualenv/python3.5.3/lib/python3.5/site-packages/pymysql/connections.py", line 706, in __init__
    self.connect()
  File "/home/travis/virtualenv/python3.5.3/lib/python3.5/site-packages/pymysql/connections.py", line 932, in connect
    self._request_authentication()
  File "/home/travis/virtualenv/python3.5.3/lib/python3.5/site-packages/pymysql/connections.py", line 1152, in _request_authentication
    auth_packet = self._read_packet()
  File "/home/travis/virtualenv/python3.5.3/lib/python3.5/site-packages/pymysql/connections.py", line 1014, in _read_packet
    packet.check_error()
  File "/home/travis/virtualenv/python3.5.3/lib/python3.5/site-packages/pymysql/connections.py", line 393, in check_error
    err.raise_mysql_exception(self._data)
  File "/home/travis/virtualenv/python3.5.3/lib/python3.5/site-packages/pymysql/err.py", line 107, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.OperationalError: (1040, 'Too many connections')
```

I'm going to see if wrapping the close function in a try/except for the odd times it does fail will resolve this while also properly closing our connections.

As a reminder, this is the exception we are getting, which should be caught be the if statement in inquisit.bounceInquisitionDbHandle():
```
Traceback (most recent call last):
  File "/home/josh/Code/Inquisition/inquisition.py", line 173, in <module>
    main()
  File "/home/josh/Code/Inquisition/inquisition.py", line 150, in main
    anatomize.startAnatomizer()
  File "/home/josh/Code/Inquisition/lib/anatomize/Anatomize.py", line 135, in startAnatomizer
    self.bounceInquisitionDbConnection()
  File "/home/josh/Code/Inquisition/lib/inquisit/Inquisit.py", line 115, in bounceInquisitionDbConnection
    self.inquisitionDbHandle.close()
  File "/usr/local/lib/python3.5/dist-packages/pymysql/connections.py", line 734, in close
    sock.close()
AttributeError: 'NoneType' object has no attribute 'close'
Traceback (most recent call last):
  File "/home/josh/Code/Inquisition/inquisition.py", line 173, in <module>
    main()
  File "/home/josh/Code/Inquisition/inquisition.py", line 150, in main
    anatomize.startAnatomizer()
  File "/home/josh/Code/Inquisition/lib/anatomize/Anatomize.py", line 135, in startAnatomizer
    self.bounceInquisitionDbConnection()
  File "/home/josh/Code/Inquisition/lib/inquisit/Inquisit.py", line 115, in bounceInquisitionDbConnection
    self.inquisitionDbHandle.close()
  File "/usr/local/lib/python3.5/dist-packages/pymysql/connections.py", line 734, in close
    sock.close()
AttributeError: 'NoneType' object has no attribute 'close'
```
```
        # check if connection already exists
        if self.inquisitionDbHandle:
            self.inquisitionDbHandle.close()
```

**Comment by magneticstain** [comment]
Seems we still have some issues, even when closing. This time instead of time_wait, we have close_wait, at least on my laptop:
```
tcp        0      0 127.0.0.1:40232         127.0.0.1:3306          ESTABLISHED 7811/python3    
tcp        0      0 127.0.0.1:40264         127.0.0.1:3306          ESTABLISHED 7821/python3    
tcp        0      0 127.0.0.1:51102         127.0.0.1:6379          ESTABLISHED 7821/python3    
tcp        0     11 127.0.0.1:40262         127.0.0.1:3306          ESTABLISHED 7820/python3    
tcp        0      0 127.0.0.1:40228         127.0.0.1:3306          ESTABLISHED 7811/python3    
tcp        1      0 127.0.0.1:40236         127.0.0.1:3306          CLOSE_WAIT  7811/python3    
tcp        0      0 127.0.0.1:51106         127.0.0.1:6379          ESTABLISHED 7819/python3    
tcp        0      0 127.0.0.1:51100         127.0.0.1:6379          ESTABLISHED 7820/python3    
tcp        1      0 127.0.0.1:40240         127.0.0.1:3306          CLOSE_WAIT  7811/python3    
tcp        1      0 127.0.0.1:40244         127.0.0.1:3306          CLOSE_WAIT  7811/python3    
tcp        0      0 127.0.0.1:40256         127.0.0.1:3306          ESTABLISHED 7817/python3    
tcp        0      0 127.0.0.1:40252         127.0.0.1:3306          ESTABLISHED 7815/python3    
tcp        0      0 127.0.0.1:40230         127.0.0.1:3306          ESTABLISHED 7811/python3    
tcp        0      0 127.0.0.1:40248         127.0.0.1:3306          ESTABLISHED 7814/python3    
tcp        0      0 127.0.0.1:51088         127.0.0.1:6379          ESTABLISHED 7814/python3    
tcp        0      0 127.0.0.1:40234         127.0.0.1:3306          ESTABLISHED 7811/python3    
tcp        0      0 127.0.0.1:40260         127.0.0.1:3306          ESTABLISHED 7819/python3    
tcp        0      0 127.0.0.1:51084         127.0.0.1:6379          ESTABLISHED 7811/python3    
tcp        0      0 127.0.0.1:51092         127.0.0.1:6379          ESTABLISHED 7815/python3    
tcp        0      0 127.0.0.1:40238         127.0.0.1:3306          ESTABLISHED 7811/python3    
tcp        0      0 127.0.0.1:40246         127.0.0.1:3306          ESTABLISHED 7811/python3    
tcp        0      0 127.0.0.1:40242         127.0.0.1:3306          ESTABLISHED 7811/python3    
######################################
```

Will see how this performs in travis CI.

**Comment by magneticstain** [comment]
This properly allows the closing of connections on Travis CI. Since it fixes that issue, while offering little difference on my dev box, we will keep this fix in place until Issue #67 is completed in the next release.
